### ISSUE TITLE
ci(tests): split the suite into fast and heavy lanes — 85% of runtime sits in 21% of the tests

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/xtremax#90]
## Problem / Request

The suite is 1142 tests and every push and every PR runs **all of them, serially, with coverage, across a 2-version Python matrix**. There are no markers, no parallelism, and no separation between millisecond-scale primitive assertions and multi-second Monte-Carlo samplers. CI is slow enough that it stops being a fast feedback signal on a PR.

We should split the suite into a **fast lane** (runs on every PR) and a **heavy lane** (runs on merge to `main`, on release tags, and on demand), mirroring the split already in production in the sibling `gaussx` repo.

## Measurements

Local box: 16 vCPU Linux, Python 3.13, `uv run pytest`, `main` @ `43c8f76`. The box had unrelated jobs running, so treat absolute numbers as *upper* bounds and the ratios as the signal.

| Configuration | Wall | Notes |
|---|---|---|
| **`uv run pytest -v` (today's CI, serial + coverage)** | **535 s (8 m 55 s)** | ×2 matrix jobs per event |
| Full suite, `-n 4`, no coverage | 187 s (3 m 07 s) | `-n 4` = GitHub-hosted runner vCPU count |
| Full suite, `-n 8`, no coverage | 151 s (2 m 31 s) | |
| **Fast lane only (`< 1.5 s` tests), `-n 4`** | **71 s** | 901 tests |
| Heavy lane only (`>= 1.5 s` tests), `-n 4` | 158 s | 241 tests |

Lanes were simulated with explicit `--deselect` lists built from the durations data, so these are real measured runs, not projections. Coverage overhead was measured separately at **+23 %** on `test_oracles.py` (24.5 s → 30.2 s).

**Proposed PR cost: ~71 s, down from ~535 s — about 7.5× faster.**

### Where the time goes

From `--durations=0` on an `-n 8` run (per-test absolute times there are inflated by worker contention; the *distribution* is the point):

| Bucket | Tests | Share of total test time |
|---|---|---|
| `>= 5 s` | 44 | 28.8 % |
| `>= 2 s` | 184 | 72.9 % |
| `>= 1.5 s` | 240 | **85.0 %** |
| `< 1.5 s` | 902 | 15.0 % |

**21 % of the tests account for 85 % of the runtime.** By subtree, `tests/point_processes/` is 69 % of it.

That split is confirmed independently by running each lane on its own at `-n 4` (rows 5 and 6 of the table above): the 241 tests over the threshold take **158 s**, the other 901 take **71 s**. One fifth of the suite costs 2.2× what the rest costs.

Slowest individual tests, from the serial + coverage run (no cross-worker inflation):

```
19.98s  test_primitives_nonstationary.py::TestNonStationary::test_reverse_mode_grad_matches_finite_difference
 7.75s  test_primitives_nonstationary.py::TestNonStationary::test_non_leading_time_axis
 7.20s  test_oracles.py::TestFloat64Isolation::test_float64_roundtrips_and_deep_tail
 5.78s  test_distributions.py::TestBijectToSupport::test_nuts_latent_site_initializes[<lambda>0]
 5.40s  test_distributions.py::TestGEVD::test_mean_excess_varies_with_threshold_at_gumbel_limit
 5.38s  point_processes/test_property_suite.py::TestMarkedGradients::test_filter_grad_finite
 5.31s  point_processes/test_distributions.py::TestHppDistribution::test_mcmc_recovers_rate
 5.30s  point_processes/test_primitives_ipp_spatial.py::...::test_intensity_recovery_by_subdomain_count
 5.24s  point_processes/test_diagnostics.py::TestKsStatisticExp1::test_hpp_samples_pass_ks
 5.12s  point_processes/test_operators.py::TestLiveLambdaAndIntegrated::test_sample_uses_live_lambda_max
```

### Root cause: XLA compilation, not numerics

The suite is **compile-bound**. Almost nothing here is FLOP-heavy — the arrays are tiny (a few thousand elements). What costs money is tracing and lowering a fresh XLA executable per distinct call signature, and `jax.lax.scan` / `fori_loop` / `while_loop` samplers lower to XLA even without an explicit `jit`.

Two independent pieces of evidence:

**1. Within a parametrized group, only the first case is slow.** Everything after it reuses the compiled kernel (times from the `-n 8` run):

```
3.11s  test_oracles.py::TestGEVPrimitiveOracles::test_cdf[-0.5]     <- first case
0.01s  test_oracles.py::TestGEVPrimitiveOracles::test_cdf[-0.2]
0.01s  test_oracles.py::TestGEVPrimitiveOracles::test_cdf[0.0]
0.01s  ... (5 more, all 0.01s)
```

**2. A warm JAX persistent compilation cache cuts runtime by ~68 %.** Same 43 tests, same box, cache dir the only difference:

```
JAX_COMPILATION_CACHE_DIR=... pytest tests/point_processes/test_primitives_hpp.py \
                                     tests/point_processes/test_primitives_ipp.py
  cold: 36.48 s
  warm: 11.52 s     (-68 %)
```

The cache was **2.2 MB / 288 entries** for those 43 tests — i.e. tens of MB for the whole suite, comfortably within an `actions/cache` entry.

This also means a naive "mark everything over 1.5 s as slow" is the *wrong* rule. In `test_oracles.py`, `test_primitives.py`, and `test_gradients.py` the tests over the threshold are just whichever case paid the first compile — marking them `slow` moves the compile into the heavy lane instead of removing it, and the fast lane would immediately re-pay it. Those three files are genuinely cheap in aggregate — 219 tests / 39 s, 209 / 52 s, 123 / 50 s at `-n 8`, i.e. under 0.2 s per test amortized. **Markers should follow behaviour (does this test draw samples / run inference / do a statistical check?), not a stopwatch threshold.**

### Secondary cost: `import xtremax` takes 5.7 s

`python -X importtime -c "import xtremax"` = **5.7 s**, of which:

- **1.9 s — `sklearn`**, pulled in eagerly by `xtremax.extraction.quantile_regression`. scikit-learn is an *optional* `[threshold]` extra, and `extraction/__init__.py` already guards the re-export with `try/except ImportError` — but when the extra *is* installed (as it is in CI) every import of `xtremax` pays for it.
- 1.4 s — `jax` / `jax.numpy` (irreducible)
- 0.9 s — `numpyro.distributions`
- 0.3 s — `scipy.signal`, from `simulations.temporal` (`lfilter`)

Serially that is a one-off 5.7 s. Under `-n N` it is paid **per worker**, which is why total CPU rises from 178 s to 293 s when the fast lane goes from serial to `-n 4`. Worth fixing on both counts.

## Proposed lane split

`gaussx` already runs exactly this and it works; port it rather than inventing a second convention.

### `pyproject.toml`

```toml
[dependency-groups]
dev = [
    "pytest>=9.0.3",
    "pytest-cov>=7.1.0",
    "pytest-xdist>=3.5",      # new
    "pre-commit>=4.5.1",
]

[tool.pytest.ini_options]
testpaths = ["tests"]
# Coverage moves OUT of addopts and into the workflows that want it, so a
# local `pytest` / `pytest -k ...` is not silently instrumented.
markers = [
    "slow: individually expensive tests (>~1.5s: samplers, Monte-Carlo, jit+grad sweeps)",
    "integration: end-to-end workflows (MCMC/SVI fits through numpyro)",
]
```

### `.github/workflows/ci.yml`

- **PR** → fast lane only: Python 3.12, `-n auto`, `-m "not slow and not integration"`, no coverage. ~1 min.
- **Push to `main`** → fast lane on the full 3.12/3.13 matrix, with coverage + Codecov (unchanged reporting).
- Add a `concurrency` group with `cancel-in-progress` for PRs so force-pushes do not queue up stale runs.

### `.github/workflows/tests-extended.yml` (new)

The heavy lane. The requirement here is "runs on merge and version bumps", so it triggers on more than `gaussx`'s dispatch-only version does:

```yaml
on:
  push:
    branches: [main]          # every merge
    tags: ["v*"]              # release-please version bumps
  schedule:
    - cron: "0 5 * * 1"       # Monday safety net
  workflow_dispatch:
    inputs:
      suite: {type: choice, default: heavy, options: [heavy, full]}
```

running `uv run pytest -v -n auto -m "slow or integration"` (or the full suite for `suite=full`), on the 3.12/3.13 matrix.

## Which tests go in which lane

Share of each file's tests that exceed 1.5 s, as measured:

| File | slow / total |
|---|---|
| `point_processes/test_operators_extensions.py` | 9/10 (90 %) |
| `point_processes/test_distributions_spatial.py` | 7/8 (88 %) |
| `point_processes/test_operators_spatiotemporal.py` | 11/14 (79 %) |
| `point_processes/test_integration_spatiotemporal.py` | 3/4 (75 %) |
| `point_processes/test_distributions_spatiotemporal.py` | 8/11 (73 %) |
| `point_processes/test_primitives_ipp_spatiotemporal.py` | 7/10 (70 %) |
| `point_processes/test_property_suite.py` | 10/15 (67 %) |
| `point_processes/test_primitives_ipp_spatial.py` | 6/9 (67 %) |
| `point_processes/test_integration_spatial.py` | 4/6 (67 %) |
| `point_processes/test_epic39_consolidation.py` | 8/13 (62 %) |
| `point_processes/test_distributions_extensions.py` | 5/8 (62 %) |
| `test_primitives_nonstationary.py` | 13/24 (54 %) |
| `point_processes/test_epic38_regressions.py` | 13/27 (48 %) |
| `point_processes/test_review_regressions.py` | 14/32 (44 %) |
| `point_processes/test_operators.py` | 9/22 (41 %) |
| `test_distributions.py` | 19/105 (18 %) |
| `test_gradients.py` | 8/123 (7 %) |
| `test_primitives.py` | 11/209 (5 %) |
| `test_oracles.py` | 8/219 (4 %) |
| `test_extraction.py`, `test_simulations.py`, `test_quantile_regression.py`, `test_xtremax.py` | **0** |

Marking rules, in priority order:

1. **`@pytest.mark.integration`** — anything that runs inference end-to-end:
   - `point_processes/test_distributions.py::TestHppDistribution::test_mcmc_recovers_rate` (NUTS, 200 warmup + 400 samples)
   - `test_distributions.py::TestBijectToSupport::test_nuts_latent_site_initializes`
   - `test_oracles.py::TestFloat64Isolation::test_float64_roundtrips_and_deep_tail` (spawns a subprocess that re-imports `xtremax` under `JAX_ENABLE_X64=1` — ~7 s, of which ~6 s is the import)
2. **`@pytest.mark.slow`** — sampler and Monte-Carlo tests: anything named `*sample*`, `*empirical*`, `*_ks*`, `*round_trip*`, `*statistics*`, plus the `filter_grad` / finite-difference gradient sweeps. This cluster alone is ~378 s (36 % of the suite).
3. **Module-level `pytestmark = pytest.mark.slow`** for the eleven files above 60 %, rather than 240 individual decorators. The handful of sub-1.5 s tests in those modules ride along into the heavy lane; that costs the fast lane nothing and keeps the marking maintainable. Files in the 40–60 % band (`test_primitives_nonstationary.py`, `test_epic38_regressions.py`, `test_review_regressions.py`, `test_operators.py`) get per-test decorators instead.
4. **Do not mark** the first-compile artifacts in `test_oracles.py` / `test_primitives.py` / `test_gradients.py`. Those files stay entirely in the fast lane.

Rules 1–3 should land as a mechanical pass, then be validated against a fresh `--durations=0` run rather than trusted blind — the marker set is a snapshot of today's timings and will drift.

## Additional speedups, ranked by payoff / effort

1. **JAX persistent compilation cache in CI** *(config only, ~68 % measured on a subset)* — set `JAX_COMPILATION_CACHE_DIR` and restore/save it with `actions/cache` keyed on `uv.lock` + jax version. Biggest single win available without touching a single test. Also worth exporting locally so `make test` benefits.
2. **Drop coverage from the PR lane** *(+23 % measured)* — keep it on the `main` push job, which is what feeds Codecov anyway. Also move it out of `addopts` so ad-hoc local runs are not instrumented.
3. **`pytest-xdist -n auto`** *(full suite: 187 s at `-n 4`, 151 s at `-n 8`)* — the suite is embarrassingly parallel. Note the scaling is sub-linear because each worker re-pays the 5.7 s import and re-compiles the kernels it needs, which is exactly what items 1 and 4 attack; total CPU for the fast lane rises from 178 s serial to 293 s at `-n 4` even as wall time drops. Worth comparing `--dist loadfile` against the default `load` — grouping a module on one worker should improve in-process compile-cache reuse.
4. **Lazy `sklearn` re-export** *(−1.9 s per process, ×N workers)* — replace the eager `try: from ... import` in `extraction/__init__.py` with a PEP 562 module `__getattr__`, so `import xtremax` does not drag in scikit-learn. Same for `scipy.signal` in `simulations/temporal.py` (−0.3 s).
5. **Collapse scalar ξ-sweeps into vectorized cases.** `XI_GRID`/`XI_BAND` are parametrized as Python scalars, so each case is a separate trace. `tests/test_gradients.py::TestGEVLogProbGradients::test_grad_wrt_x_loc_scale` is 6.28 s for the first case and ~0.3 s each for the other 17 — the ~0.3 s is per-call tracing that never amortizes because nothing is `jit`-wrapped. Running the grid as one array-valued assertion would cut this to a single trace. Trade-off: coarser failure attribution, so apply it only where the parametrize ID is not doing diagnostic work.
6. **Session-scoped fixtures for shared setup.** The whole suite defines 6 `@pytest.fixture`s, none with a `scope=`; plain helper methods like `TestNonStationary._fields()` and `_ipp()` rebuild identical inputs on every test. Minor next to the compile cost, but free.
7. **Trim Monte-Carlo sample counts on the fast lane.** `test_oracles.py::TestSampleStatistics` draws 100 000 and 20 000 samples for moment/KS checks; `test_simulations.py` uses `n_years=200`. These belong in the heavy lane at full size rather than being shrunk.

## Implementation Steps

- [ ] Add `pytest-xdist` to the `dev` group; add `markers` to `[tool.pytest.ini_options]`; move `--cov*` out of `addopts`
- [ ] Apply `@pytest.mark.integration` to the three inference/subprocess tests
- [ ] Apply module-level `pytestmark = pytest.mark.slow` to the eleven `point_processes/` modules above 60 %
- [ ] Apply per-test `@pytest.mark.slow` in the 40–60 % band and to the sampler/Monte-Carlo tests in `test_distributions.py`
- [ ] Rewrite `.github/workflows/ci.yml` into `test-fast` (PR) + `test-main` (push) jobs; add the `concurrency` group
- [ ] Add `.github/workflows/tests-extended.yml` (push to `main`, `v*` tags, weekly cron, `workflow_dispatch`)
- [ ] Add `JAX_COMPILATION_CACHE_DIR` + `actions/cache` to both workflows
- [ ] Make `sklearn` and `scipy.signal` imports lazy (`__getattr__`)
- [ ] Add `make test-fast` / `make test-heavy`; update `CLAUDE.md`, `AGENTS.md`, `docs/contributing.md`
- [ ] Re-run `--durations=0` and confirm the marker set still matches the measured split

## Definition of Done

- [ ] `pytest -m "not slow and not integration"` completes in **< 2 min** on a 4-core runner
- [ ] Heavy lane runs on merge to `main`, on `v*` tags, weekly, and on `workflow_dispatch`
- [ ] Every test is reachable from exactly one of the two lanes; `pytest` with no `-m` still runs everything
- [ ] Coverage reporting to Codecov is unchanged in content
- [ ] `make test && make lint && make typecheck` green

## Testing

- [ ] `pytest -m "not slow and not integration" --collect-only` and `pytest -m "slow or integration" --collect-only` sum to 1142
- [ ] Heavy lane green on both 3.12 and 3.13

## Documentation

- [ ] Document the two lanes and the markers in `CONTRIBUTING.md` / `docs/contributing.md`
- [ ] Add `make test-fast` / `make test-heavy` targets alongside the existing `test` / `test-cov`

## Relationships

- Related: `gaussx` — reference implementation of this exact split (`.github/workflows/ci.yml` + `tests-extended.yml`, `slow`/`integration` markers)
- Related: #75 (already anticipated `@pytest.mark.slow` for the simulator statistical tests; never landed)

**Comment by jejjohnson** [comment]
Implemented in #92 — markers, lane split, `tests-extended.yml`, and the JAX compilation cache. Measured on a real runner: the PR gate went from ~4m 53s (full suite + coverage on a 2-version matrix) to **59 s** single-job, with all checks green.

Two follow-ups from this issue are deliberately **not** in #92 and remain open here:

- Lazy `sklearn` / `scipy.signal` imports (speedup item 4) — changes `src/` import behaviour, so it wants its own PR.
- Collapsing the scalar ξ-sweeps into vectorized cases (item 5) — trades failure-attribution granularity for compile count, worth discussing before doing.
